File: examples/typescript/src/shared-data.ts
// Shared Data Example
// --------------------
// Demonstrates: collaborative state with save, get, key$, and optimistic concurrency
//
// Run: npm run shared-data
// Requires: Starfish server running at ws://localhost:8080/starfish
//
// Two clients share a counter and a config object, demonstrating
// replace, counter.add, and merge operations with version tracking.

import { StarfishClient } from "@starfish/client";
import WebSocket from "ws";

const SERVER_URL = "ws://localhost:8080/starfish";

function createClient(name: string) {
  return new StarfishClient({
    server: SERVER_URL,
    ws: (url) => new WebSocket(url) as any,
    client: { name },
  });
}

async function main() {
  const clientA = createClient("Client-A");
  const clientB = createClient("Client-B");

  await clientA.connect();
  await clientB.connect();
  await clientA.join("shared-data-demo");
  await clientB.join("shared-data-demo");
  console.log("Both clients joined session.\n");

  // Watch all data changes from Client-B's perspective
  clientB.changed$.subscribe((result) => {
    console.log(`[Client-B] Data changed: ${result.key} = ${JSON.stringify(result.data)} (v${result.version})`);
  });

  await sleep(200);

  // --- Replace operation: set a value ---
  console.log("Client-A sets 'config' to { theme: 'dark', fontSize: 14 }");
  const r1 = await clientA.save({
    key: "config",
    scope: "session",
    op: "replace",
    data: { theme: "dark", fontSize: 14 },
  });
  console.log(`  Saved at version ${r1.version}\n`);
  await sleep(300);

  // --- Merge operation: partial update ---
  console.log("Client-B merges { fontSize: 18 } into 'config'");
  const r2 = await clientB.save({
    key: "config",
    scope: "session",
    op: "merge",
    data: { fontSize: 18 },
  });
  console.log(`  Merged at version ${r2.version}\n`);
  await sleep(300);

  // --- Get operation: read current value ---
  const current = await clientA.get({ key: "config", scope: "session" });
  console.log(`Client-A reads 'config': ${JSON.stringify(current.data)} (v${current.version})\n`);

  // --- Counter operation ---
  console.log("Client-A initializes 'score' to 0");
  await clientA.save({ key: "score", scope: "session", op: "replace", data: 0 });
  await sleep(200);

  console.log("Client-A increments 'score' by 10");
  await clientA.save({ key: "score", scope: "session", op: "counter.add", data: 10 });
  await sleep(200);

  console.log("Client-B increments 'score' by 5");
  await clientB.save({ key: "score", scope: "session", op: "counter.add", data: 5 });
  await sleep(200);

  const score = await clientA.get({ key: "score", scope: "session" });
  console.log(`\nFinal score: ${score.data} (v${score.version})`);

  // --- Watch specific key ---
  console.log("\nClient-A watches 'status' key:");
  clientA.key$("status").subscribe((result) => {
    console.log(`  status updated to: ${JSON.stringify(result.data)}`);
  });

  await clientB.save({ key: "status", scope: "session", op: "replace", data: "ready" });
  await sleep(200);
  await clientB.save({ key: "status", scope: "session", op: "replace", data: "running" });
  await sleep(200);

  // Clean up
  await clientA.disconnect();
  await clientB.disconnect();
  console.log("\nDone.");
}

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: examples/python/shared_data.py
# Shared Data Example
# --------------------
# Demonstrates: collaborative state with save, get, data_key_stream, and optimistic concurrency
#
# Run: python shared_data.py
# Requires: Starfish server running at ws://localhost:8080/starfish
#
# Two clients share a counter and a config object, demonstrating
# replace, counter.add, and merge operations with version tracking.

import asyncio
import json
import os

from starfish import StarfishClient, StarfishClientOptions, ClientIdentity, SaveOptions

SERVER_URL = os.environ.get("STARFISH_SERVER_URL", "ws://localhost:8080/starfish")


def create_client(name: str) -> StarfishClient:
    return StarfishClient(
        StarfishClientOptions(server=SERVER_URL, client=ClientIdentity(name=name))
    )


async def main():
    client_a = create_client("Client-A")
    client_b = create_client("Client-B")

    await client_a.connect()
    await client_b.connect()
    await client_a.join("shared-data-demo")
    await client_b.join("shared-data-demo")
    print("Both clients joined session.\n")

    # Watch all data changes from Client-B's perspective
    client_b.data_changed.subscribe(
        lambda r: print(f"[Client-B] Data changed: {r.key} = {json.dumps(r.data)} (v{r.version})")
    )

    await asyncio.sleep(0.2)

    # --- Replace operation: set a value ---
    print("Client-A sets 'config' to { theme: 'dark', fontSize: 14 }")
    r1 = await client_a.save(
        SaveOptions(key="config", scope="session", op="replace", data={"theme": "dark", "fontSize": 14})
    )
    print(f"  Saved at version {r1.version}\n")
    await asyncio.sleep(0.3)

    # --- Merge operation: partial update ---
    print("Client-B merges { fontSize: 18 } into 'config'")
    r2 = await client_b.save(
        SaveOptions(key="config", scope="session", op="merge", data={"fontSize": 18})
    )
    print(f"  Merged at version {r2.version}\n")
    await asyncio.sleep(0.3)

    # --- Get operation: read current value ---
    current = await client_a.get("config", "session")
    print(f"Client-A reads 'config': {json.dumps(current.data)} (v{current.version})\n")

    # --- Counter operation ---
    print("Client-A initializes 'score' to 0")
    await client_a.save(SaveOptions(key="score", scope="session", op="replace", data=0))
    await asyncio.sleep(0.2)

    print("Client-A increments 'score' by 10")
    await client_a.save(SaveOptions(key="score", scope="session", op="counter.add", data=10))
    await asyncio.sleep(0.2)

    print("Client-B increments 'score' by 5")
    await client_b.save(SaveOptions(key="score", scope="session", op="counter.add", data=5))
    await asyncio.sleep(0.2)

    score = await client_a.get("score", "session")
    print(f"\nFinal score: {score.data} (v{score.version})")

    # --- Watch specific key ---
    print("\nClient-A watches 'status' key:")
    client_a.data_key_stream("status").subscribe(
        lambda r: print(f"  status updated to: {json.dumps(r.data)}")
    )

    await client_b.save(SaveOptions(key="status", scope="session", op="replace", data="ready"))
    await asyncio.sleep(0.2)
    await client_b.save(SaveOptions(key="status", scope="session", op="replace", data="running"))
    await asyncio.sleep(0.2)

    # Clean up
    await client_a.disconnect()
    await client_b.disconnect()
    print("\nDone.")


if __name__ == "__main__":
    asyncio.run(main())
